### PR TITLE
feat: dump config both after loaded and ready

### PR DESCRIPTION
**Pull request by dead-horse**
##### Checklist

- [x] `npm test` passes
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message follows commit guidelines

### CODE REVIEW

File: lib/egg.js [hunk 1]
```diff
@@ -40,7 +40,7 @@ class EggApplication extends EggCore {
      */
     this.messenger = new Messenger();
 
-    this.dumpConfig();
+    this.beforeStart(() => this.dumpConfig());
```
Comment (fengmk2) [inline lib/egg.js]: Should dump after all beforeStart hooks executed.

Maybe `afterStart()` hook. cc @popomore

File: test/fixtures/apps/demo/app.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+'use strict';
+
+module.exports = app => {
+  app.config.dynamicConfig = true;
```
Comment (popomore) [inline test/fixtures/apps/demo/app.js]: The test need change config asynchronously

File: lib/egg.js (hunk 1 above)
Comment (popomore) [inline lib/egg.js]: So this should move after `this.loader.load()` that can get all config after load.

Comment (popomore) [inline lib/egg.js]: afterStart is ready

Comment (dead-horse) [inline lib/egg.js]: but sometimes we want to check the config file to figure out why server is not ready

Comment (fengmk2) [inline lib/egg.js]: @dead-horse dump config in both hooks. :)

Comment (popomore) [inline lib/egg.js]: Or dump twice with different name

Comment (atian25) [inline lib/egg.js]: `server is not ready` mean `start fail` ?  so, no need to dump twice

**Comment by dead-horse** [comment]
dump config both after loaded and after ready.

File: test/fixtures/apps/dumpconfig-app/app.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+const sleep = require('ko-sleep');
```
Comment (fengmk2) [inline test/fixtures/apps/dumpconfig-app/app.js]: 这个文件是多余的

Comment (dead-horse) [inline test/fixtures/apps/dumpconfig-app/app.js]: 删掉了

**Pull request merged**